File: project/core/module/rest-service/src/main/java/vdi/service/rest/health/DependencySource.java
package vdi.service.rest.health;

import org.slf4j.LoggerFactory;
import org.veupathdb.lib.container.jaxrs.health.Dependency;
import org.veupathdb.lib.container.jaxrs.health.ExternalDependency;
import vdi.core.health.RemoteDependencies;

import java.util.Iterator;

public class DependencySource implements org.veupathdb.lib.container.jaxrs.health.DependencySource {
  @Override
  public Iterator<Dependency> iterator() {
    return new Iterator<>() {
      private final Iterator<vdi.core.health.Dependency> raw = RemoteDependencies.INSTANCE.iterator();

      @Override
      public boolean hasNext() {
        return raw.hasNext();
      }

      @Override
      public Dependency next() {
        return new RemoteDependency(raw.next());
      }
    };
  }

  private static class RemoteDependency extends ExternalDependency {
    private final vdi.core.health.Dependency raw;

    public RemoteDependency(vdi.core.health.Dependency raw) {
      super(raw.getName());
      this.raw = raw;
    }

    @Override
    public TestResult test() {
      LoggerFactory.getLogger(getClass()).info("checking health for external dependency {}", name);

      var status = switch(raw.checkStatus()) {
        case Ok -> Status.ONLINE;
        case NotOk -> Status.OFFLINE;
        case Unknown -> Status.UNKNOWN;
      };

      var reachable = (status == Status.UNKNOWN && pinger.isReachable(raw.getHost(), raw.getPortAsInt()))
        || status == Status.ONLINE;

      return new TestResult(this, reachable, status, raw.getExtra());
    }

    @Override
    public void close() {}
  }
}

File: project/core/module/rest-service/src/main/kotlin/vdi/service/rest/health/DependencySource.kt
package vdi.service.rest.health

import org.veupathdb.lib.container.jaxrs.health.Dependency
import org.veupathdb.lib.container.jaxrs.health.Dependency.Status
import org.veupathdb.lib.container.jaxrs.health.Dependency.TestResult
import org.veupathdb.lib.container.jaxrs.health.DependencySource
import org.veupathdb.lib.container.jaxrs.health.ExternalDependency
import vdi.core.health.RemoteDependencies
import vdi.logging.logger
import vdi.core.health.Dependency as VDep

internal class DependencySource : DependencySource {
  override fun iterator(): MutableIterator<Dependency> {
    return object : MutableIterator<Dependency> {
      private val raw = RemoteDependencies.iterator()

      override fun hasNext() = raw.hasNext()

      override fun next(): Dependency = RemoteDependency(raw.next())

      override fun remove() {
        throw UnsupportedOperationException()
      }
    }
  }
}

private fun VDep.Status.toCoreStatus() = when (this) {
  VDep.Status.OK -> Status.ONLINE
  VDep.Status.NOT_OK -> Status.OFFLINE
  VDep.Status.UNKNOWN -> Status.UNKNOWN
}

private class RemoteDependency(private val raw: VDep) : ExternalDependency(raw.name) {
  override fun close() {}
  override fun test(): TestResult {
    logger().info("Checking health for external dependency {}", this.name)
    val status = raw.checkStatus().toCoreStatus()
    val reachable = (status == Status.UNKNOWN && pinger.isReachable(raw.host, raw.port))
      || status == Status.ONLINE
    return TestResult(this, reachable, status, raw.extraFields)
  }
}
